### PR TITLE
Implement 'uet_mr_regv'

### PR DESCRIPTION
**Pull request by svyatoslavnik**
Function needed for `fi_mr_regv`.

Tested with `fi_mr_test` fabtest, no tests are available so far at uet-ref-prov level

Feature tested internally with all fixes from other PRs included. Since this branch is based off main, PDS tests are expected to fail until https://github.com/ultraethernet/uet-ref-prov/pull/52 is merged. Once merged with https://github.com/ultraethernet/uet-ref-prov/pull/52 into main, it should work as expected.

### CODE REVIEW

File: uet_api.c [hunk 1]
```diff
@@ -4527,6 +4527,83 @@ int uet_mr_reg(uet_domain_handle_t domain_handle, void *buf, size_t len,
...
+		if (rc != FI_SUCCESS)
+			return rc;
+	}
+
+	/* init memory region descriptor */
+	mr_desc = &uet_dom->mr_desc[mr_index];
+	memset(mr_desc, 0, sizeof(struct uet_mr_desc));
+	mr_desc->state = UET_MR_DESC_STATE_DISABLED_REG;
+	mr_desc->uet_dom = uet_dom;
+	mr_desc->buf_desc.type = UET_MR_BUF_TYPE_IOV;
```
Comment (alexb-kt) [inline uet_api.c]: Looks like `nic_rawsock_mr_reg()` and `nic_xdp_mr_reg()` (called via `uet_nic_mr_reg()` from `uet_mr_enable()`) both balk on types that are not `_CONTIG`; so presumably they'd need patching too.

(or is there any other way to exercise this code path?)

File: uet_api.c [hunk 2]
```diff
@@ -4527,6 +4527,83 @@ int uet_mr_reg(uet_domain_handle_t domain_handle, void *buf, size_t len,
 	return FI_SUCCESS;
 }
 
+int uet_mr_regv(uet_domain_handle_t domain_handle, const struct iovec *iov,
```
Comment (alexb-kt) [inline uet_api.c:4472]: Most of the function seems a duplicate of `uet_mr_reg()`, the common portion should be factored out.

`uet_mr_reg()` could even be a special case that constructs a size=1 `iovec` and then calls `uet_mr_regv()`.

File: uet_api.c [hunk 3]
```diff
@@ -4527,6 +4527,83 @@ int uet_mr_reg(uet_domain_handle_t domain_handle, void *buf, size_t len,
...
+			return rc;
+	}
+
+	/* init memory region descriptor */
+	mr_desc = &uet_dom->mr_desc[mr_index];
+	memset(mr_desc, 0, sizeof(struct uet_mr_desc));
+	mr_desc->state = UET_MR_DESC_STATE_DISABLED_REG;
+	mr_desc->uet_dom = uet_dom;
+	mr_desc->buf_desc.type = UET_MR_BUF_TYPE_IOV;
+	mr_desc->buf_desc.iov.iov = iov;
```
Comment (alexb-kt) [inline uet_api.c]: I don't know if there's an API contract for this (I don't see one in the `fi_mr_regv()` man page), but it seems surprising to hold a reference to the `struct iovec *` from the caller -- e.g. they could be sitting on the stack.

File: uet_api.c (hunk 2 above)
Comment (svyatoslavnik) [reply uet_api.c:4472]: Thank you for this comment, @alexb-kt. Yes, you are right. We have already done this with sendv, recvv, and have it in our plans. We may update this PR if we implement it before the PR is accepted.

Our primary goal now is to move forward and meet fabtest requirements. We can and will polish the code later.

File: uet_api.c (hunk 1 above)
Comment (svyatoslavnik) [reply uet_api.c]: Hi @alexb-kt, thank you for your comment. We find it relevant, and more — we don't see any need to descend to Network functionality (NIC_SHIM) for registering and enabling memory regions (there isn't even a proper dereg function; it only returns FI_SUCCESS). We believe this should completely stay at the transport layer, and we have it in our plans.

It seems like this deserves a separate discussion, maybe a quick call or posting an issue. What do you think?

File: uet_api.c (hunk 2 above)
Comment (svyatoslavnik) [reply uet_api.c:4472]: Hi @alexb-kt , we added a commit, and refactoring is now  in this PR

File: uet_api.c (hunk 3 above)
Comment (svyatoslavnik) [reply uet_api.c]: Hi @alexb-kt, while implementing this feature, we aimed to follow the same style as used for providers other than UET. Other providers follow this approach as well.

Comment (alexb-kt) [reply uet_api.c]: Not sure (I'm certainly not an authority); I see many providers have `mr_iov_limit` of 1, but one that doesn't is `psm3` and it copies each iov item: https://github.com/ofiwg/libfabric/blob/f03fe010b869f110e488dc3d4e8dda08870341cf/prov/psm3/src/psmx3_mr.c#L423

Others might end up in the generic `ofi_mr_*` functions, which also have a copy function: https://github.com/ofiwg/libfabric/blob/f03fe010b869f110e488dc3d4e8dda08870341cf/prov/util/src/util_mr_map.c#L63

Comment (alexb-kt) [reply uet_api.c]: On that note, we (uet) also have a `mr_iov_limit` of 1 :) https://github.com/ultraethernet/uet-libfabric/blob/4c5cd1b2212778f0e5626b208f310a1a716d8f85/prov/uet/src/uet_init.c#L85

File: uet_api.c (hunk 1 above)
Comment (alexb-kt) [reply uet_api.c]: Maintainers might know (@insanum ?) -- I would imagine it's there to emulate having hardware that needs to know about the MRs as well.

File: uet_api.c (hunk 3 above)
Comment (svyatoslavnik) [reply uet_api.c]: Yes, @alexb-kt You are right. After more accurate review we fixed this. Now our implementation also do copy.

It is still a question do we need IOV with count more than one or not. Thank you for pointing out to the limitation set by uet-libfabric. On the other hand in the uet-ref-prov it is limited to 8 in the `uet_api_private.h`:

```
#define UET_IOV_LIMIT     8
#define UET_RMA_IOV_LIMIT 8
#if UET_IOV_LIMIT > UET_RMA_IOV_LIMIT
	#define UET_IOV_LIMIT_MAX UET_IOV_LIMIT
#else
	#define UET_IOV_LIMIT_MAX UET_RMA_IOV_LIMIT
#endif
```

@insanum what are your thoughts on this?

File: uet_api.c (hunk 1 above)
Comment (insanum) [reply uet_api.c]: That is correct. The NIC shim calls are there in the case (to mimic) anything the hardware might need to do when an MR is allocated.

File: uet_api.c (hunk 3 above)
Comment (insanum) [reply uet_api.c]: Note sure if we can limit to just one iov. Seems limiting and I suppose there are applications that would like more. Probably best to support more than one.

File: uet_api.c (hunk 1 above)
Comment (svyatoslavnik) [reply uet_api.c]: Yes, sure. And we see PDS/SES layers as hardware too.
Do you?

**Pull request merged**